### PR TITLE
Modify AES-CTR to not reinit after being keyed

### PR DESCRIPTION
**Pull request by ColtonWilley**

### CODE REVIEW

**Comment by SparkiDev** [comment]
Is there an explicit test for this changes?

File: test/test_cipher.c [hunk 1]
```diff
@@ -599,7 +599,22 @@ int test_aes_ctr_leftover_data_regression(ENGINE *e, void *data)
         }
     }
 
-    /* Try the other way, now. Encrypt with wolfEngine, decrypt with wolfSSL. */
+    /* Try the other way, now. Encrypt with wolfEngine, decrypt with openSSL.
+     * The EVP_CIPHER_CTX remembers any engine it was loaded with, meaning we
+     * need to reset the ctxs before reuse or the decCtx will still pick up
+     * wolfEngine */
+    if (encCtx != NULL)
```
Comment (bandi13) [inline test/test_cipher.c:606]: You should put these in their respective `if (err == 0)` blocks.

**Comment by ColtonWilley** [comment]
> Is there an explicit test for this changes?

This re-init flow already exists in test_aes_ctr_leftover_data_regression(). It used to work before commit https://github.com/wolfSSL/wolfssl/commit/6b8280f6635cee566d28ada175b6739fe8440444, but that commit changes wc_AesInit() to set rounds to zero, thus causing a new but valid error from wc_AesEncrypt() because rounds is zero.

**Comment by kaleb-himes** [comment]
> > Is there an explicit test for this changes?
> 
> This re-init flow already exists in test_aes_ctr_leftover_data_regression(). It used to work before commit [wolfSSL/wolfssl@6b8280f](https://github.com/wolfSSL/wolfssl/commit/6b8280f6635cee566d28ada175b6739fe8440444), but that commit changes wc_AesInit() to set rounds to zero, thus causing a new but valid error from wc_AesEncrypt() because rounds is zero.



I may have misunderstood  but I think @SparkiDev was asking if an explicit test will be added to cover the case.

The issue identified came from the call sequence: init-with-key then init-iv where the init-iv wiped out the rounds set by init-with-key

Possibly adding a test in tests/api.c that checks the rounds after a double init call? (Pseudo code below)

```
int test_aes_ctr_multi_init()
{
     EXPECT_DECLS;
     int rounds_expected = #rounds expected to be there as a result of the first call init'ing with key;
     /* Test invalid sequence */
     ExpectIntEQ(we_aes_ctr_init(with key set), WOLFSSL_SUCCESS);
     ExpectIntEQ(we_aes_ctr_init(with all the NULL params and only the IV passed in), rounds_expected);
     return EXPECT_RESULT();
}
```

**Comment by ColtonWilley** [comment]
> Possibly adding a test in tests/api.c that checks the rounds after a double init call?

If such tests were to be added, that should be a direct test of the wolfcrypt APIs, I dont think wolfengine tests are a good place for verifying AES context internals. We could add another test with just a double init, but it would ultimately be a copy of the existing test code. If the consensus is that we absolutely want new test code I can add some, but I think it would be a copy of existing code that doesnt actually add anything new in terms of testing.

**Comment by kaleb-himes** [comment]
Fair enough.

**Comment by bandi13** [comment]
The failure in `gcc, clang, scan-build, fsanitize and disable test(s)` can be ignored. It was a test that accidentally ran on this PR. It has been disabled.

**Review by douzzer — APPROVED** [review]
LGTM.  as discussed, more fixes will be needed for more leaks around `wc_AesInit()`, and possibly for more over-inits.

**Pull request merged**